### PR TITLE
fixes #584, #583, #581

### LINKED ISSUE #584: Want async shutdown test for websocket
As this has been a challenge for us in the past, @wirebirdlabs has written a test case for async shutdown to check for races against issues elsewhere in the library.

### PR DESCRIPTION
**Pull request by JackDunaway**

### CODE REVIEW

File: src/transports/ws/sws.c [hunk 1]
```diff
@@ -678,7 +686,76 @@ static void nn_sws_validate_utf8_chunk (struct nn_sws *self)
...
+    pos = self->inmsg_current_chunk_buf;
+
+    /*  Peer did not provide a Close Code, so choose our own here. */
+    if (len == 0) {
+        nn_sws_fail_conn (self, NN_SWS_CLOSE_NORMAL, "");
+        return;
+    }
+
+    /*  If the payload is not even long enough for the required 2-octet
+        Close Code, the connection should have been failed upstream. */
```
Comment (gdamore) [inline src/transports/ws/sws.c]: What does "upstream" mean in this context?

File: src/transports/ws/sws.c [hunk 2]
```diff
@@ -678,7 +686,76 @@ static void nn_sws_validate_utf8_chunk (struct nn_sws *self)
...
+    nn_assert (len >= NN_SWS_CLOSE_CODE_LEN);
+    len -= NN_SWS_CLOSE_CODE_LEN;
+    pos += NN_SWS_CLOSE_CODE_LEN;
+
+    /*  As per RFC 6455 7.1.6, the Close Reason following the Close Code
```
Comment (gdamore) [inline src/transports/ws/sws.c]: While this is required for generation, do we care about actually checking this?  What happens if we just ignore the close code whether well formed or not?  (Be generous in what you accept.... )

File: src/transports/ws/sws.c [hunk 3]
```diff
@@ -678,7 +686,76 @@ static void nn_sws_validate_utf8_chunk (struct nn_sws *self)
...
+        close_code == NN_SWS_CLOSE_ERR_EXTENSION ||
+        close_code == NN_SWS_CLOSE_ERR_SERVER ||
+        (close_code >= 3000 && close_code <= 3999) ||
+        (close_code >= 4000 && close_code <= 4999)) {
+        /*  Repeat close code, per RFC 6455 7.4.1 and 7.4.2 */
+        nn_sws_fail_conn (self, (int) close_code, "");
+    }
+    else {
+        nn_sws_fail_conn (self, NN_SWS_CLOSE_ERR_PROTO,
+            "Unrecognized close code.");
```
Comment (gdamore) [inline src/transports/ws/sws.c]: TBH, I'm not sure what value passing the close code here gives.  How is this useful to the application?

File: src/transports/ws/sws.c [hunk 4]
```diff
@@ -759,7 +835,7 @@ static int nn_sws_fail_conn (struct nn_sws *self, int code, char *reason)
         nn_fsm_raise (&self->fsm, &self->done, NN_SWS_RETURN_CLOSE_HANDSHAKE);
     }
 
-    return 0;
+    return;
```
Comment (gdamore) [inline src/transports/ws/sws.c]: Return statement here not needed.

File: src/transports/ws/sws.c [hunk 5]
```diff
@@ -1200,13 +1272,25 @@ static void nn_sws_handler (struct nn_fsm *self, int src, int type,
                             sws->inmsg_current_chunk_buf = sws->inmsg_control;
                         }
                         else {
-                            sws->inmsg_chunks++;
                             sws->inmsg_total_size += sws->inmsg_current_chunk_len;
+                            /*  Protect non-control messages against the
```
Comment (gdamore) [inline src/transports/ws/sws.c]: Yay.  This is a critical stopper bug you are fixing. :-)

File: src/transports/ws/sws.c [hunk 6]
```diff
@@ -1298,8 +1379,19 @@ static void nn_sws_handler (struct nn_fsm *self, int src, int type,
...
                         sws->inmsg_total_size += sws->inmsg_current_chunk_len;
+                        /*  Protect non-control messages against the
+                            NN_RCVMAXSIZE threshold; control messages already
+                            have a small pre-allocated buffer, and therefore
+                            are not subject to this limit. */
+                        nn_pipebase_getopt (&sws->pipebase, NN_SOL_SOCKET,
+                            NN_RCVMAXSIZE, &opt, &opt_sz);
+                        if (opt >= 0 && sws->inmsg_total_size > (size_t) opt) {
+                            nn_sws_fail_conn (sws, NN_SWS_CLOSE_ERR_TOOBIG,
+                                "Message larger than application allows.");
```
Comment (gdamore) [inline src/transports/ws/sws.c]: This message, if exposed to users, could be confusing.  "Message larger than NN_RCVMAXSIZE" might be clearer.  Or just "message size exceeds limit"?

File: tests/ws_async_shutdown.c [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        if (nn_errno () == ETIMEDOUT) {
+            continue;
+        }
+        break;
+    }
+    /*  Socket is expected to be closed by caller.  */
+    errno_assert (nn_errno () == EBADF);
+}
+
+int main ()
```
Comment (gdamore) [inline tests/ws_async_shutdown.c]: I'd prefer to see argc, argv arguments (even if unused) to main().

File: src/transports/ws/bws.c [hunk 8]
```diff
@@ -314,6 +344,71 @@ static void nn_bws_handler (struct nn_fsm *self, int src, int type,
         }
 
 /******************************************************************************/
+/*  CLOSING_USOCK state.                                                      */
```
Comment (gdamore) [inline src/transports/ws/bws.c]: Just a little note.  I hate the use of enclosing asterisks in block comments.  It becomes a pain to maintain them, and it looks horrible if ever used with a non-monospaced font (I know don't do that!) I used to use the same comment style when I was young, but in my years since I've learned that they add little value -- the ones on the right side are especially annoying to have to maintain.  Usually just a single multiline comment with left aligned asterisks is sufficient to catch attention -- especially if surrounded by whitespace.

Comment (gdamore) [inline src/transports/ws/bws.c]: That said, I think Martin used these block comments extensively elsewhere in the code.  They still annoy me, and I celebrate a little when I can remove them. :-)

Comment (JackDunaway) [inline src/transports/ws/bws.c]: Agree here. I'll leave the format for now for conformance, then we can come back later and change them all in bulk.

File: src/transports/ws/sws.c (hunk 1 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: Upstream logic in the currently-processing message. Not upstream in the sense of "further up the chain" on a block diagram of the system.

File: src/transports/ws/sws.c (hunk 2 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: The Close Reason itself it optional, but if it exists, it **MUST BE** valid UTF-8. If it is not, RFC 6455 states that we (local socket) can immediately fail the connection without best-effort response. If it's well-formed, RFC 6455 states we must give a best-effort response. I see where you're coming from, but relaxing the spec here in this manner would be incorrect.

Comment (gdamore) [inline src/transports/ws/sws.c]: So it would be wrong to give a best effort response to this, even if the reason was not valid UTF-8?

File: src/transports/ws/sws.c (hunk 3 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: It's just RFC 6455 conformance. A well-behaved server (like this logic) should provide best-effort hints as to why it's unexpectedly closing a connection when it does, based on the malconformance mode of the client.

File: src/transports/ws/sws.c (hunk 1 above)
Comment (gdamore) [inline src/transports/ws/sws.c]: So lets change the wording here, because its unclear.  Perhaps say "should already have been failed."

File: src/transports/ws/sws.c (hunk 4 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: IIRC, it's something that pedantic compilers prefer; either way, it doesn't hurt to be explicit. Do you have strong opionions otherwise? Else, we'll keep it for now.

File: src/transports/ws/sws.c (hunk 6 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: Good suggestion; will change.

File: tests/ws_async_shutdown.c (hunk 7 above)
Comment (JackDunaway) [inline tests/ws_async_shutdown.c]: Even further, this test needs the full-blown upgrade made by @Snaipe to dynamically retrieve its port. (This test predates that upgrade.) Stay tuned ...

File: src/transports/ws/sws.c (hunk 1 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: Will change.

File: src/transports/ws/sws.c (hunk 2 above)
Comment (JackDunaway) [inline src/transports/ws/sws.c]: No, since a best-effort response is still given for all other failure modes. It's a paltry effort, but it tries at least once prior to closing the connection.

Comment (JackDunaway) [inline src/transports/ws/sws.c]: Also -- it's very helpful to see how all this nanomsg websocket code interacts with the Autobahn Test Suite. It's simple to run A/B tests on question about behavior like this one, since it does an excellent job of probing all corner cases.

Comment (gdamore) [inline src/transports/ws/sws.c]: Ok, lets check this at some point -- I'd rather not have the burden of validating UTF-8-ishness if we can skip it.  Seems to be of very little merit here, IMO.

File: src/transports/ws/sws.c (hunk 3 above)
Comment (gdamore) [inline src/transports/ws/sws.c]: I see, this response is sent back to the far end, so its for peer consumption rather than app consumption.  Right?

File: src/transports/ws/sws.c (hunk 4 above)
Comment (gdamore) [inline src/transports/ws/sws.c]: No "pedantic" compiler that I'm aware of "prefers" an explicit return at the end of a void function.  If there is a returned value, then that's different.   (Implicit return values of zero are what compilers bitch about.)  I'd prefer to remove it, actually.
